clang/lib/Sema/SemaDecl.cpp

```diff
@@ -16318,11 +16318,15 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
     checkTypeSupport(FD->getType(), FD->getLocation(), FD);
 
   // Handle free functions.
-  if (LangOpts.SYCLIsDevice && FD->hasAttr<SYCLDeviceAttr>() && Body &&
-      (FD->getTemplatedKind() == FunctionDecl::TK_NonTemplate ||
-       FD->getTemplatedKind() ==
-           FunctionDecl::TK_FunctionTemplateSpecialization))
-    SYCL().ProcessFreeFunction(FD);
+  if (FD) {
+    // Free functions cannot be member functions so skip those.
+    const auto *MD = dyn_cast<CXXMethodDecl>(FD);
+    if (LangOpts.SYCLIsDevice && !MD && Body &&
```

Comment:
Based on your tests below and explanation in PR comments, it sounds like this issue does not have anything to do with `constexpr` values. The issue here is we have non-instantiated functions/methods being checked when they shouldn't be.  Can you try replacing this check with `LangOpts.SYCLIsDevice && Body && FD->isDependentContext()` without the `TemplatedKind` checks. You also do not need the `if (FD)` since L16305 above checks this already.

I think you should also replace the `hasDependentExpr` checks you added in SemaSYCL to asserts instead because we should not be having dependent arguments at that point.

Reply:
Slight correction, that should be `!FD->isDependentContext()` -- basically, if the function isn't dependent, it's safe to check the arguments; if the function IS dependent somehow, we need to wait to do those checks until the function has been instantiated (regardless of whether it's a member function or free function).

Reply:
The `hasDependentExpr `check I added was a check before calling `getAttributeNameValuePairs` because that would assert if all expressions in the attribute were not constexprs. So there is already an assert there. Perhaps an assert before calling `getAttributeNameValuePairs` is not needed?

Making the other change in SemaDecl.cpp was fine. The tests all pass.

Reply:
Ok. You can remove all that the code you added then if one already exists in  `getAttributeNameValuePairs `

Reply:
Done.